sensors/baro/ms5611.c

```diff
@@ -0,0 +1,280 @@
+		fprintf(stderr, "ms5611: adc read failed\n");
+		return -1;
+	}
+
+	/* If ADC is read too early then it returns 0 */
+	if (val[0] == 0 && val[1] == 0 && val[2] == 0) {
+		return -1;
+	}
+
+	*res = ((uint32_t)val[0] << 16) | ((uint32_t)val[1] << 8) | ((uint32_t)val[2]);
+
+	return 0;
+}
+
+
+static void ms5611_publishthr(void *data)
+{
+	sensor_info_t *info = (sensor_info_t *)data;
+	ms5611_ctx_t *ctx = info->ctx;
+	int32_t temp, press;
+	time_t now;
+
+	/* breaking naming convention on purpose to match datasheet notation 1:1 */
+	uint32_t D1 = 0, D2 = 0;
+	int32_t dT;
+	int64_t off, sens;
+
+	/* Initialize temperature with too big delay for certain response */
+	while (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_4096), OSR_4096_SLEEP * 2, &D2) < 0)
+		;
+
+	/* Translate temperature */
+	dT = (int32_t)D2 - (((int32_t)ctx->prom[5]) << 8);           /* dT = D2 - C5 * 2^8 */
+	temp = 2000 + (((int64_t)dT * (int64_t)ctx->prom[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
+
+	while (1) {
+		/* Read pressure */
+		if (ms5611_measure(ctx, (CMD_CVRT_PRESS | CVRT_OSR_4096), OSR_4096_SLEEP, &D1) < 0) {
+			continue;
+		}
+
+		gettime(&now, NULL);
+
+		/* Read temperature. Does not repeat failed read - fall back to old temperature */
+		if (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_1024), OSR_1024_SLEEP, &D2) == 0) {
+			/* Translate temperature */
+			dT = (int32_t)D2 - (((int32_t)ctx->prom[5]) << 8);           /* dT = D2 - C5 * 2^8 */
+			temp = 2000 + (((int64_t)dT * (int64_t)ctx->prom[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
+		}
+
+		/* Translate press */
+		off = ((int64_t)ctx->prom[2] << 16) + (((int64_t)dT * (int64_t)ctx->prom[4]) >> 7);  /* OFF = C2 * 2^16 + (C4 * dT) / 2^7 */
+		sens = ((int64_t)ctx->prom[1] << 15) + (((int64_t)dT * (int64_t)ctx->prom[3]) >> 8); /* SENS = C1 * 2^15 + (C4 * dT) / 2^8 */
+		press = ((((int64_t)D1 * sens) >> 21) - off) >> 15;                                  /* P = ((D1 * SENS) / 2^21 - OFF) / 2^15 */
+
+		/* MS5611 operating ranges: temperature from -40C to +80C, and pressure: 45000Pa, 110000 Pa */
+		if (press > 45000 && press < 120000 && temp > -4000 && temp < 8500) {
+			ctx->evtBaro.timestamp = now;
+			ctx->evtBaro.baro.pressure = press;
+			ctx->evtBaro.baro.temp = (temp / 100) + 273;
+
+			sensors_publish(info->id, &ctx->evtBaro);
+		}
+	}
+
+	return;
+}
+
+
+static int ms5611_hwSetup(ms5611_ctx_t *ctx)
+{
+	uint8_t cmd, i;
+
+	/* Reset ms5611 sequence */
+	cmd = CMD_RESET;
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), NULL, 0, 0) < 0) {
+		fprintf(stderr, "ms5611: failed to reset device\n");
+		return -1;
+	}
+	usleep(RESET_SLEEP);
+
+	/* PROM reading */
+	for (i = 0; i < PROM_SIZE; i++) {
+		cmd = CMD_READ_PROM | (i << 1); /* LSB of prom address is always 0 */
+		if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &ctx->prom[i], sizeof(ctx->prom[i]), sizeof(cmd)) < 0) {
```

Comment:
Doesn't this device handle address autoincrement on register read?

Reply:
Unfortunately it does not. It is not stated in datasheet. Also it is not used in AN520 application note (there is an example code of ms5611 driver). Even if it works without any datasheet mention I would stick to loop reading.